source/MaterialXGenShader/ShaderNode.h

```diff
@@ -246,6 +248,7 @@ class MX_GENSHADER_API ShaderPort : public std::enable_shared_from_this<ShaderPo
   protected:
     ShaderNode* _node;
     TypeDesc _type;
+    ConstStructMemberDescVecPtr _structMembers;
```

Comment:
I'm curious why you need to store the struct members description on every port? Would it not be possible to find these from the TypeDesc?

Reply:
The TypeDesc object itself doesn't store the structure members (because we want to keep that small), but instead the struct members are stored in the TypeDescStorage object, but there is a need to access the struct members in MaterialXRenderGLSL/MSL modules, which don't have access to a GenContext.  The compromise here is to pass them along on the port downstream - most of the time this will just be a `nullptr` if the port isn't a struct..

Reply:
Ah I see! Would it be possible to instead store this pointer on the actual TypeDesc objects?

I suspect the reason you didn't do that in the first place (and instead used the `_structIndex`) was to keep the TypeDesc objects at 64-bits. But now we have crossed that barrier anyway and the TypeDesc objects are 80 bits. So if we change the `uint16_t _structIndex` to a pointer holding the struct description we get 128 bits instead (two full machine words), which probably is equal or even better for performance compared to 80.

Reply:
You're completely right - I was trying not to increase the size of TypeDesc, but if we can just store the pointer in the TypeDesc itself, then it feels more natural to me, and might completely do away with the secondary storage.  

Let me take another pass and see how things shake out in that direction - thanks Niklas.

Reply:
I think it came out a lot cleaner - and certainly simplified the data storage. A number of other things might get even simpler if we could use a similar approach to store the name string with a pointer inside TypeDesc, instead of in the separate map... Ultimately I like the OIIO::ustring idea - but I think that needs to be a separate PR to work through what it means to add the dependency.

Reply:
That's great to see it turned out well.

I agree a lot more could be simplified by having access to the type name from inside the TypeDesc. In fact I think that is a requirement, since it doesn't feel right to store the type name in various places in addition to the TypeDesc itself. 

I couldn't resist to take a stab at this, so I've made a draft change list for storing a pointer to the name on the TypeDesc. This removed a lot of the issues with having to pass the context around just to find the type name.

That commit has been added here now. To be honest I didn't mean to push this commit to your branch - I ment to push this to my own fork of your branch, but that failed miserably, and apparently it was pushed here instead :) Sorry about that, but maybe that's a good thing as we can now review the merged result. And if this turns out to not be a good path forward we can always revert the commit.

Reply:
No worries - team work makes the dream work!  Just ping me either here - or on slack - once you're done pushing to the PR - so we don't clobber each other.